external-crates/move/crates/move-compiler/src/typing/deprecations.rs

```diff
@@ -0,0 +1,765 @@
+            .map(|deprecation| (deprecation, position))
+    }
+
+    fn deprecated_constant(
+        &self,
+        mident: &ModuleIdent,
+        name: &ConstantName,
+    ) -> Option<(Deprecation, AttributePosition)> {
+        self.deprecated_constants
+            .get(&(*mident, *name))
+            .or_else(|| self.deprecated_module(mident))
+            .cloned()
+            .map(|deprecation| (deprecation, AttributePosition::Constant))
+    }
+
+    fn deprecated_function(
+        &self,
+        mident: &ModuleIdent,
+        name: &FunctionName,
+    ) -> Option<(Deprecation, AttributePosition)> {
+        self.deprecated_functions
+            .get(&(*mident, *name))
+            .or_else(|| self.deprecated_module(mident))
+            .cloned()
+            .map(|deprecation| (deprecation, AttributePosition::Function))
+    }
+
+    // Look up the deprecation information for a given module. If the module is deprecated, and we
+    // are _not_ calling from within the deprecated module, return the deprecation information. If
+    // we are calling from within the deprecated module, return None. This prevents us for emitting
+    // warnings for things like the call to `f` in this module.
+    // ```text
+    // #[deprecated]
+    // module 0x42::m {
+    //   fun f() { }
+    //   fun g() { f(); } // call to f will _not_ emit a deprecated warning since this is within 0x42::m
+    // }
+    // ```
+    // Note that we will always emit a deprecated warning for any calls to function that is
+    // annotated as deprecated, regardless of whether the call is within the module or not:
+    // ```text
+    // module 0x42::m {
+    //   #[deprecated]
+    //   fun f() { }
+    //   fun g() { f(); } // call to f will emit a deprecated warning
+    // }
+    // ```
+    //
+    fn deprecated_module(&self, mident: &ModuleIdent) -> Option<&Deprecation> {
+        // Dont' register a deprecation warning if the module is already deprecated and we are
+        // calling from within the deprecated module.
+        if Some(*mident) == self.current_mident {
+            return None;
+        }
+
+        self.deprecated_modules.get(mident)
+    }
+
+    // Register a deprecation warning for a given deprecation and location. This collects all
+    // `loc`s within a given named context. We will then then merge them into a single warning at
+    // the end based on this bucketing.
+    fn register_deprecation_warning(
+        &mut self,
+        (deprecation, member_type): (Deprecation, AttributePosition),
+        loc: Loc,
+    ) {
+        let mident = self
+            .current_mident
+            .expect("ICE: current module should always be set when visiting deprecations");
+        let name = self
+            .current_named_context
+            .expect("ICE: current named context should always be set when visiting deprecations");
+        let deprecation_id = DeprecationId {
+            mident,
+            name,
+            deprecation,
```

Comment:
I think this is unnecessary? You just need to know the attribute position I think?
You can then fetch the `Deprecation` when you render the warnings

Reply:
I believe this is still necessary -- we use this to bucket warnings to emit into the `(named_context, deprecation)` pairs. So it felt right to do this like this (to me at least)